stan/math/rev/core/vari.hpp

```diff
@@ -247,6 +259,17 @@ class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
     ChainableStack::instance_->var_stack_.push_back(this);
   }
 
+  template <typename S, require_convertible_t<S&, T>* = nullptr,
+            require_arena_matrix_vt<std::is_arithmetic, S>* = nullptr>
+  explicit vari_value(const S& x)
```

Comment:
S should be a const arena_matrix for this to happen, otherwise the memory might accidentally get changed after the function is called.

It makes sense to modify some `arena_matrix` variables after construction and in an elementwise way, so we should only take the memory if it is const.

Do we need `is_arithmetic` here?

Reply:
> S should be a const arena_matrix for this to happen, otherwise the memory might accidentally get changed after the function is called.

> It makes sense to modify some arena_matrix variables after construction and in an elementwise way, so we should only take the memory if it is const.

Do you mean the arena matrix could point to new memory as in the pointer to memory will change in the `arena_matrix`? tbh I'm not sure if that's a big thing to worry about since this is a copy op. 

> Do we need is_arithmetic here?

Yes since we treat the arithmetic data as the values

```cpp
      : val_mem_(x.data()),
        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
            x.size())),
```

We would need a different specialization for `arena_matrix<Matrix<var>>` that would require hard copies of both the `val_` and `adj_` mem

Reply:
> Do you mean the arena matrix could point to new memory

I mean:

```cpp
arena_matrix<MatrixXd> memory(100, 100);
arena_matrix<var_value<MatrixXd>> result = memory;

memory = 1.0; // result changes too!
```

I think it should only steal the pointer to the values if memory is `const`.

Reply:
I still think const goes here.

Reply:
> I think it should only steal the pointer to the values if memory is const.

If it's const idt we can take the memory because you can't have a non-const reference to a const object

```
const double a = 5;
double& b = a; // fails
```

I think stealing that memory makes sense. Like when we copy a var we just copy over the `*vi_` pointer so the same thing happens like

```cpp
  using stan::math::var;
  var A(5);
  var B = A;
  B.vi_->adj_ = 4;
  std::cout << "Adjoint of A: " << A.vi_->adj_;
  // Adjoint of A: 4
```

Reply:
This would be a nasty bug if it actually happens anywhere. @bbbales2 Can you think of a realistic example where arena_matrix is used to construct a var_value and than changed? Or another way - where a var_value is changed and the arena_matrix it is constructed from should not? After some thinking about this I can't think of any such example.

Reply:
> After some thinking about this I can't think of any such example.

@t4c1 yeah I agree with what you're saying. Had a talk with @SteveBronder this morning too and he convinced me that we shouldn't worry about this.